src/Npgsql.DependencyInjection/NpgsqlServiceCollectionExtensions.cs

```diff
@@ -28,19 +28,22 @@ public static class NpgsqlServiceCollectionExtensions
     /// The lifetime with which to register the <see cref="NpgsqlDataSource" /> service in the container.
     /// Defaults to <see cref="ServiceLifetime.Singleton" />.
     /// </param>
+    /// <param name="serviceKey">The <see cref="ServiceDescriptor.ServiceKey"/> of the data source.</param>
     /// <returns>The same service collection so that multiple calls can be chained.</returns>
     public static IServiceCollection AddNpgsqlDataSource(
         this IServiceCollection serviceCollection,
         string connectionString,
         Action<NpgsqlDataSourceBuilder> dataSourceBuilderAction,
         ServiceLifetime connectionLifetime = ServiceLifetime.Transient,
-        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
-        => AddNpgsqlDataSourceCore(serviceCollection, connectionString, dataSourceBuilderAction, connectionLifetime, dataSourceLifetime);
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton,
```

Comment:
One recommendation would be to use https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/baseline-version-validator (this exact scenario is listed in the example for why this tool is helpful).

Reply:
> This is a binary breaking change. Are you OK with making it?

I am, these methods are rarely used outside app projects so binary breaks are not *that* impactful.

We could add back the original methods and move the serviceKey argument to the back of the required argument list but it'd be a fairly noisy overload set. @eerhardt do you think it's worth it?

> One recommendation would be to use https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/baseline-version-validator (this exact scenario is listed in the example for why this tool is helpful).

We're using https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md across Npgsql projects but it seems like it's not enabled in this one.

Reply:
@roji just pointed out we could reintroduce the original overloads and make all its optional parameters required. At that point we are able to keep serviceKey at the end of the new overload. 

With that in place we could optionally mark the original method `EditorBrowsable(Never)` as well, to get rid of the noise...

If I have a moment before we release 8.0 I'll probably do something like that.

Reply:
My personal opinion is that `AddKeyedNpgsqlDataSource` makes it more understandable what this does. If I just happen to pass in a `key` into this method, I can no longer get my NpgsqlDataSource from DI the "typical" way of just GetService (or a ctor param). Now I need to call GetKeyedService.

> I am, these methods are rarely used outside app projects so binary breaks are not that impactful.

Just note that this would break .NET Aspire if it was built earlier using the 7.x versions of Npgsql.DependencyInjection. .NET Aspire calls this method: https://github.com/dotnet/aspire/blob/8a7b625e0da646d12b272aa26e69c0abe4ac7439/src/Components/Aspire.Npgsql/AspirePostgreSqlNpgsqlExtensions.cs#L109. .NET Aspire's component library `Aspire.Npgsql` would no longer load if an application updated Npgsql.DependencyInjection to this new version. So I wouldn't underestimate how impactful this kind of change can be.  (Again, it is the canonical scenario listed for why the package validation tool was written.)

Reply:
> My personal opinion is that AddKeyedNpgsqlDataSource makes it more understandable what this does. If I just happen to pass in a key into this method, I can no longer get my NpgsqlDataSource from DI the "typical" way of just GetService (or a ctor param). Now I need to call GetKeyedService.

Right - but that's true of the other registration "aspect" - the lifetime.. i.e. if someone accidentally specifies Scoped instead of Singleton, the application will start misbehaving very badly. The service key doesn't seem very different from that conceptually - it's another aspect of the service registration; (it's true that it's an object rather than something typed, but IMHO it's reasonable to expect people to understand what a parameter means before passing in an argument...

We actually did consider putting Keyed in the method name (see https://github.com/npgsql/npgsql/pull/5387#discussion_r1390426896), but there already are enough Npgsql-specific "facets" that this seemed to become a bit unmanageable. For example, a data source can be slim (for less size) and also multi-host (a special kind of data source for use with multiple database servers, failover etc.), which would bring us to AddKeyedMultiHostNpgsqlSlimDataSourceCore.

Does that make sense?

Reply:
We are aware of the changes that would be binary breaking. The primary cause here is not having the public api analyzer enabled and acting too quickly to get an 8.0 release out. 

So in any case thanks for bringing it up, it might certainly save some pain.

>Again, it is the canonical scenario listed for why the package validation tool was written

The public api analyzer may not spell out the rationale of why certain changes are binary breaks. However any change that needs you to mark an api as *REMOVED* is essentially binary breaking. It would have brought sufficient attention to it for us to consider it more carefully. Once I do make the change in this area I'll be sure to set up the public api analyzer.